src/aapns/api.py

```diff
@@ -1,14 +1,111 @@
-from typing import Optional
+from __future__ import annotations
 
-import attr
+import abc
+import asyncio
+import os
+from dataclasses import dataclass, replace
+from tempfile import TemporaryDirectory
+from typing import Optional
 
 from . import config, errors, models
 from .pool import Pool, Request, create_ssl_context
 
 
-@attr.s(auto_attribs=True, frozen=True)
-class APNS:
-    server: config.Server
+class APNSBaseClient(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
```

Comment:
man you love those abstract classes a tad too much 😓

Reply:
Would you consider PEP-544 "Structural subtyping" instead?
https://www.python.org/dev/peps/pep-0544/